File: cpp/int-ai4.cpp
#include <iostream>
#include <cmath>    // For sin() and M_PI

// Function to calculate f(x) = sin(x)
double f(double x) {
    return sin(x);
}

// Trapezoidal Rule Implementation
double trapezoidalRule(double a, double b, int n) {
    double h = (b - a) / n;  // Step size
    double area = 0.0;
    double x = a;

    // Sum area of each trapezoid
    for (int i = 0; i < n; ++i) {
        area += 0.5 * h * (f(x) + f(x + h));
        x += h;
    }

    return area;
}

// Function to compute the area under sin(x) between a and b, considering absolute values
double computeAbsoluteArea(double a, double b, int n) {
    double h = (b - a) / n;  // Step size
    double area = 0.0;
    double x = a;

    // Sum absolute areas under the curve
    for (int i = 0; i < n; ++i) {
        area += 0.5 * h * (fabs(f(x)) + fabs(f(x + h)));
        x += h;
    }

    return area;
}

int main() {
    double a = -M_PI;
    double b = 2.0 / 3.0 * M_PI;
    int n = 1000;  // Number of intervals (higher means more accuracy)

    double area = trapezoidalRule(a, b, n);
    double absArea = computeAbsoluteArea(a, b, n);

    std::cout << "Approximate integral of sin(x) from -pi to 2/3*pi: " << area << std::endl;
    std::cout << "Total area between sin(x) and the x-axis from -pi to 2/3*pi: " << absArea << std::endl;

    return 0;
}

File: python/int-ai4.py
###Python Code Using Numpy for Trapezoidal Integration
import numpy as np

def trapezoidal_rule(f, a, b, n):
    x = np.linspace(a, b, n + 1)
    y = f(x)
    h = (b - a) / n
    return h * (np.sum(y) - 0.5 * (y[0] + y[-1]))

# Define the function and limits
f = np.abs(np.sin)  # Taking the absolute value of sin(x)
a = -np.pi
b = 2 * np.pi / 3
n = 10000  # Number of intervals

# Calculate the area using the trapezoidal rule
area = trapezoidal_rule(f, a, b, n)
print(f"The total area between sin(x) and the x-axis from -pi to 2/3*pi is: {area:.4f}")
